src/index/base.cpp

```diff
@@ -101,8 +113,16 @@ void BaseIndexNotifications::blockConnected(const interfaces::BlockInfo& block)
     // best block is not updated here before sync, but this has been
     // longstanding behavior since syncing was introduced in #13033 so care
     // should be taken if changing m_best_block_index semantics.
-    if (m_index.m_synced) {
+    assert(synced == m_index.m_synced);
+    if (synced) {
         m_index.m_best_block_index = pindex;
+    } else if (m_last_locator_write_time + SYNC_LOCATOR_WRITE_INTERVAL < current_time || m_index.m_interrupt) {
```

Comment:
commit 1819bdb7cb0634dad2e07daccfe7e8a122a943d6:
I think that this changed behavior slightly because it changes the order of indexing the block (`CustomAppend`) and periodically updating the index' best block (`Commit`)  - which is better because the indexed data on disk wouldn't be in an inconsistent state temporarily. I opened #25074 to fix this on master.